Season 03/JeongIn/Week19/2281.py

```diff
@@ -0,0 +1,36 @@
+# 2281 데스노트
+# DP(메모이제이션)
+
+import sys
+sys.setrecursionlimit(10**6)
+input = sys.stdin.readline
+# INF = (999 ** 2) * 999 # 최악의 경우 999칸씩 999번 비게 됨
+INF = 10 ** 9 # 위 경우보다 메모리 4KB 더 차지하지만 4ms 단축
+
+n, m = map(int, input().split()) # n : 사람 수 , m : 노트 가로 칸 수(폭)
+names_len = [] # 이름 글자수 저장하는 리스트
+dp = [[-1] * m for _ in range(n)] # dp[i][j] 는 i번째 사람 이름을 썼을 때 총 칸수 제곱의 합
+
+
+for _ in range(n):
+    names_len.append(int(input()))
+
+def write_name(next_idx, remain):
+
+    if next_idx == n :
+        return 0
+
+    # 아직 써야할 사람이 남아있는 경우
+    if dp[next_idx][remain] < 0 :
+        next_len = names_len[next_idx]
+        ans = INF
+
+        if next_len < remain :
+            ans = min(ans, write_name(next_idx+1, remain - names_len[next_idx] - 1)) # 현재 줄에 쓸 수 있는 경우
+
+        ans = min(ans, write_name(next_idx+1, m - names_len[next_idx]) + remain **2 ) # 다음 줄로 넘겨쓸 경우
```

Comment:
어짜피 결국에는 다음줄로 넘기기 vs 현재줄에 쓰기 둘중에 제일 작은값을 골라야하는건데 현재줄에 쓸수없는 경우가 생기는거잖아요 ? . . 

그러면 INF 쓰지않고도

```suggestion
        
        ans = write_name(next_idx+1, m - names_len[next_idx]) + remain **2  # 다음 줄로 넘겨쓸 경우

        if next_len < remain :
            ans = min(ans, write_name(next_idx+1, remain - names_len[next_idx] - 1)) # 현재 줄에 쓸 수 있는 경우

```
요렇게 했어도 돼지않았을까 . . .

Reply:
오옹 맞아요..! 사실 첨에 다음 줄로 넘겨 쓰는 경우를 먼저 구하는 걸로 풀었었는데 또 이상한데서 에러가 나서 다른 방식으로 풀어보았습니다 ㅎㅎ

Reply:
좋은 답안 제시 감사합니다 헤헷